File: src/codes/forced-compatibility/py/rfc_proxy.py
from abc import ABC, abstractmethod

class Printer(ABC):
    @abstractmethod
    def print(self, text: str):
        pass

class RealPrinter(Printer):
    def print(self, text: str):
        print(f"印刷中: {text}")

class PrinterProxy(Printer):
    def __init__(self):
        self._real = RealPrinter()

    def print(self, text: str):
        print("ログ: 印刷を開始します")
        self._real.print(text)
        print("ログ: 印刷が完了しました")

# 利用例
printer = PrinterProxy()
printer.print("月次レポート")

File: src/codes/forced-compatibility/ts/rfc_proxy.ts
// インターフェース
interface Printer {
  print(text: string): void;
}

// 実際の処理を行うクラス（本体）
class RealPrinter implements Printer {
  print(text: string): void {
    console.log(`印刷中: ${text}`);
  }
}

// Proxy（アクセスを仲介）
class PrinterProxy implements Printer {
  private real: RealPrinter;

  constructor() {
    this.real = new RealPrinter();
  }

  print(text: string): void {
    console.log("ログ: 印刷を開始します");
    this.real.print(text);
    console.log("ログ: 印刷が完了しました");
  }
}

// 利用例
const printer: Printer = new PrinterProxy();
printer.print("月次レポート");
